mlir/tools/mlir-miopen-driver/mlir-miopen-driver.cpp

```diff
@@ -55,86 +56,166 @@ static cl::opt<std::string> outputFilename("o", cl::desc("Output filename"),
-
-static cl::opt<int> deviceNum(
-    "device",
-    cl::desc("Device index on which to run the kernel (only with host code)"),
-    cl::value_desc("Between 0 and number of GPUs on system. "
-                   "Omission leaves current device intact."));
-static cl::alias deviceShort("dev", cl::aliasopt(deviceNum));
-
 namespace test {
 void registerTestDialect(DialectRegistry &);
 } // namespace test
 
-static void populateTuningPipeline(PassManager &pm) {
-  pm.addPass(
-      mlir::miopen::createAffixTuningParametersPass(blockSize, gridSize));
+static LogicalResult
+parsePipeline(StringRef pipeline, llvm::SmallDenseSet<StringRef> &pipelineSet,
+              llvm::SmallDenseSet<StringRef> &pipelineOptions) {
+  size_t start = 0;
+  size_t pos = 0;
+  do {
+    pos = pipeline.find_first_of(",", start);
+    auto opt = pipeline.substr(start, pos).trim();
+    if (opt.empty()) {
+    } else if (opt == "all" || opt == "full") {
+      pipelineSet = pipelineOptions;
+    } else if (pipelineOptions.contains(opt)) {
+      pipelineSet.insert(opt);
+    } else {
+      SmallString<256> opts;
+      for (auto pipeline : pipelineOptions) {
+        if (opts.size())
+          opts += ", ";
+        opts += pipeline;
+      }
+      llvm::errs() << "Invalid pipeline: " << opt << "\n"
+                   << "   Valid options: " << opts << " or full\n";
+      return failure();
+    }
+    start = pos + 1;
+  } while (pos != StringRef::npos);
+
+  return success();
 }
 
 static LogicalResult runMLIRPasses(ModuleOp &module,
                                    mlir::PassPipelineCLParser &passPipeline) {
-  PassManager pm(module.getContext(), PassManager::Nesting::Implicit);
+  llvm::SmallDenseSet<StringRef> kernelPipelineOptions{"tuning", "gpu", "rocdl",
+                                                       "binary"};
+  llvm::SmallDenseSet<StringRef> kernelPipelineSet;
+  if (failed(parsePipeline(kernelPipeline.getValue(), kernelPipelineSet,
+                           kernelPipelineOptions))) {
+    return failure();
+  }
+
+  llvm::SmallDenseSet<StringRef> hostPipelineOptions{"partition", "highlevel",
+                                                     "xmodel"};
+  llvm::SmallDenseSet<StringRef> hostPipelineSet;
+  if (failed(parsePipeline(hostPipeline.getValue(), hostPipelineSet,
+                           hostPipelineOptions))) {
+    return failure();
+  }
+
+  // Run partitioning pipeline.
+  if (hostPipelineSet.contains("partition")) {
+    PassManager pm(module.getContext(), PassManager::Nesting::Implicit);
+    pm.addPass(tosa::createTosaPartitionPass());
+    pm.addPass(miopen::createMIOpenCloneKernelsPass());
+
+    if (failed(pm.run(module))) {
+      return failure();
+    }
+  }
+
+  // Find kernel module, defaults to top module
+  auto kernelModule =
+      module.lookupSymbol<ModuleOp>(miopen::MIOpenDialect::kKernelModuleName);
+  if (!kernelModule) {
+    kernelModule = module;
+  }
```

Comment:
Aren't we guaranteed to have a kernel module at this point? If so I'd prefer this as an assert rather than we default kernel module to module.